core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java

```diff
@@ -4882,4 +4924,113 @@ private boolean isH2DB(Connection dbConnection) throws Exception {
 
         return H2.equalsIgnoreCase(DatabaseCreator.getDatabaseType(dbConnection));
     }
+
+    /**
+     * Circuit Breaker pattern added for DB connection retrieval
+     *          Renamed original getDBConnection() method to getConnection() to minimize changes
+     *
+     * @return DB Connection
+     * @throws UserStoreException if unknown occurred while getting database connection.
+     * @throws SQLException if error occurred while retrieving database connection
+     */
+    protected Connection getDBConnection() throws UserStoreException, SQLException {
+
+        Connection dbConnection = null;
+
+        switch (jdbcConnectionCircuitBreakerState) {
+            case CIRCUIT_STATE_OPEN:
+                long circuitOpenDuration = System.currentTimeMillis() - thresholdStartTime;
+                if (circuitOpenDuration >= thresholdTimeoutInMilliseconds) {
+                    try {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Trying to obtain JDBC connection for " + getMyDomainName() +" domain" +
+                                    " when circuit breaker state: "
+                                    + jdbcConnectionCircuitBreakerState + " and circuit breaker open duration: "
+                                    + circuitOpenDuration + "ms.");
+                        }
+                        dbConnection = getConnection();
+                        jdbcConnectionCircuitBreakerState = CIRCUIT_STATE_CLOSE;
+                        thresholdStartTime = 0;
+                        break;
+                    } catch (Exception e) {
+                        log.error("Error occurred while obtaining connection", e);
+                        thresholdStartTime = System.currentTimeMillis();
```

Comment:
Are we failing only after 2? Can this be a configurable one... as this depends on the system